[Final Final Review] Add readable output format for diff tool

wjhsf opened the pull request:
Example output of `raml-toolkit diff --dir base-dir new-dir`:
```
API Changed: shopper-search/shopper-search.raml
  Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.16/SecuritySchemes/shopper-token.raml#/fragment/parameter/Authorization
    [Breaking] Parameter Added: Authorization

<snipped>

  Summary
    4 Breaking Changes
    52 Ignored Changes

API Changed: cdn-api-process-apis/cdnapi-process-api.raml
  Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2Finfo/get
    [Breaking] Display Name Changed: getZoneInfo → getZonesInfo
  Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Fgroups%2F%7BgroupId%7D/put
    [Breaking] Display Name Changed: updateWafGroupById → updateWafGroup
  Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Frules%2F%7BruleId%7D/put
    [Breaking] Display Name Changed: updateWafRuleById → updateWafRule
  Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/get
    [Breaking] Display Name Changed: getFirewallRuleById → getFirewallRule

<snipped>

  Summary
    10 Breaking Changes
    68 Ignored Changes

API Added: shopper-login/shopper-login.raml

API Removed: shopper-stores/shopper-stores.raml

API Removed: einstein-api-quick-start-guide/einstein-api-quick-start-guide.raml

Summary
  1 APIs added
  2 APIs removed
  18 APIs changed
```

Output of `raml-toolkit diff base.raml new.raml` is the same as the content under an "API Changed" section (but without the indent).

Output of `raml-toolkit --diff-only` currently is "No changes." That flag will be changed to only allow JSON output.

Currently ignored rules have no special handling and they are included in the output. Concurrent story `W-7822674` will remove "ignored" as a listable category.

TODO: Tests / docs.

wjhsf commented:
Still not 100% done, but I wanted to share the updated format:
```
APIs Changed
────────────

  File: cdn-api-process-apis/cdnapi-process-api.raml
  ──────────────────────────────────────────────────

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2Finfo/get
      Ignored Changes: 1

      Breaking Changes: 1
      ───────────────────
      Display Name Changed: getZoneInfo → getZonesInfo

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Fgroups%2F%7BgroupId%7D/put
      Breaking Changes: 1
      ───────────────────
      Display Name Changed: updateWafGroupById → updateWafGroup

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Frules%2F%7BruleId%7D/put
      Breaking Changes: 1
      ───────────────────
      Display Name Changed: updateWafRuleById → updateWafRule

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/get
      Breaking Changes: 1
      ───────────────────
      Display Name Changed: getFirewallRuleById → getFirewallRule

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
      Breaking Changes: 1
      ───────────────────
      Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
      Breaking Changes: 1
      ───────────────────
      Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
      Breaking Changes: 1
      ───────────────────
      Parameter Added: WWW-Authenticate

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/delete/500/application%2Fproblem%2Bjson/schema/ErrorResponse/property/type/scalar/type/example/default-example
      Ignored Changes: 1

APIs Added
──────────
  shopper-login/shopper-login.raml

APIs Removed
────────────
  shopper-stores/shopper-stores.raml
  einstein-api-quick-start-guide/einstein-api-quick-start-guide.raml

Summary
───────
  Changed APIs: 18
    - Breaking Changes: 564
    - Ignored Changes: 3147
  Added APIs: 1
  Removed APIs: 2
```

unandyala commented:
> Still not 100% done, but I wanted to share the updated format:
> 
> ```
> APIs Changed
> ────────────
> 
>   File: cdn-api-process-apis/cdnapi-process-api.raml
>   ──────────────────────────────────────────────────
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2Finfo/get
>       Ignored Changes: 1
> 
>       Breaking Changes: 1
>       ───────────────────
>       Display Name Changed: getZoneInfo → getZonesInfo
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Fgroups%2F%7BgroupId%7D/put
>       Breaking Changes: 1
>       ───────────────────
>       Display Name Changed: updateWafGroupById → updateWafGroup
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Frules%2F%7BruleId%7D/put
>       Breaking Changes: 1
>       ───────────────────
>       Display Name Changed: updateWafRuleById → updateWafRule
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/get
>       Breaking Changes: 1
>       ───────────────────
>       Display Name Changed: getFirewallRuleById → getFirewallRule
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
>       Breaking Changes: 1
>       ───────────────────
>       Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
>       Breaking Changes: 1
>       ───────────────────
>       Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
>       Breaking Changes: 1
>       ───────────────────
>       Parameter Added: WWW-Authenticate
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/delete/500/application%2Fproblem%2Bjson/schema/ErrorResponse/property/type/scalar/type/example/default-example
>       Ignored Changes: 1
> 
> APIs Added
> ──────────
>   shopper-login/shopper-login.raml
> 
> APIs Removed
> ────────────
>   shopper-stores/shopper-stores.raml
>   einstein-api-quick-start-guide/einstein-api-quick-start-guide.raml
> 
> Summary
> ───────
>   Changed APIs: 18
>     - Breaking Changes: 564
>     - Ignored Changes: 3147
>   Added APIs: 1
>   Removed APIs: 2
> ```

I feel we don't need summary/counts per node. We probably can have summary per API but for directory diff I feel its good enough to just have the overall summary

wjhsf commented:
New format:
```
APIs Changed
────────────

  File: orders/orders.raml
  ────────────────────────

    Node: #/web-api/https%3A%2F%2F%7BshortCode%7D.api.commercecloud.salesforce.com%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/shortCode
      [Breaking] Parameter Added: shortCode

    Node: #/web-api/https%3A%2F%2F%7BshortCode%7D.api.commercecloud.salesforce.com%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/version
      [Breaking] Parameter Added: version

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D/patch
      [Breaking] Display Name Changed: patchSitesByIdOrdersById → updateOrder

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fconfirmation-status/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdConfirmationStatus → updateOrderConfirmationStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fexport-status/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdExportStatus → updateOrderExportStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fexternal-status/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdExternalStatus → updateOrderExternalStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-status/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdPaymentStatus → updateOrderPaymentStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fshipping-status/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdShippingStatus → updateOrderShippingStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fstatus/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdStatus → updateOrderStatus

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-instruments%2F%7BpaymentInstrumentId%7D/patch
      [Breaking] Display Name Changed: patchSitesByIdOrdersByIdPaymentInstrumentsById → updateOrderPaymentInstrument

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-instruments%2F%7BpaymentInstrumentId%7D%2Ftransaction/patch
      [Breaking] Display Name Changed: patchSitesByIdOrdersByIdPaymentInstrumentsByIdTransaction → updateOrderPaymentTransaction

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fshipments%2F%7BshipmentId%7D%2Fshipping-address/put
      [Breaking] Display Name Changed: putSitesByIdOrdersByIdShipmentsByIdShippingAddress → updateOrderShippingAddress

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
      [Breaking] Parameter Added: WWW-Authenticate

    Node: #/web-api/https%3A%2F%2Fanypoint.mulesoft.com%2Fmocking%2Fapi%2Fv1%2Flinks%2F2a968e56-a214-422b-a6e2-61ed41faad5f%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/version
      [Breaking] Parameter Removed: version

    Node: #/web-api/https%3A%2F%2Fanypoint.mulesoft.com%2Fmocking%2Fapi%2Fv1%2Flinks%2F2a968e56-a214-422b-a6e2-61ed41faad5f%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/shortCode
      [Breaking] Parameter Removed: shortCode

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.8/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Removed: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.8/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
      [Breaking] Parameter Removed: WWW-Authenticate

    Summary
    ───────
      Breaking Changes: 19
      Ignored Changes: 163

File: cdn-api-process-apis/cdnapi-process-api.raml
  ──────────────────────────────────────────────────

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2Finfo/get
      [Breaking] Display Name Changed: getZoneInfo → getZonesInfo

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Fgroups%2F%7BgroupId%7D/put
      [Breaking] Display Name Changed: updateWafGroupById → updateWafGroup

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Frules%2F%7BruleId%7D/put
      [Breaking] Display Name Changed: updateWafRuleById → updateWafRule

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/get
      [Breaking] Display Name Changed: getFirewallRuleById → getFirewallRule

    Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/put
      [Breaking] Display Name Changed: updateFirewallRuleById → updateFirewallRule

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Added: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
      [Breaking] Parameter Added: WWW-Authenticate

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.10/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
      [Breaking] Parameter Removed: Authorization

    Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.10/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
      [Breaking] Parameter Removed: WWW-Authenticate

    Summary
    ───────
      Breaking Changes: 10
      Ignored Changes: 68

APIs Added
──────────
  shopper-login/shopper-login.raml

APIs Removed
────────────
  shopper-stores/shopper-stores.raml
  einstein-api-quick-start-guide/einstein-api-quick-start-guide.raml

Summary
───────
  Changed APIs: 18
    - Breaking Changes: 564
    - Ignored Changes: 3147
  Added APIs: 1
  Removed APIs: 2
```

unandyala commented:
> New format:
> 
> ```
> APIs Changed
> ────────────
> 
>   File: orders/orders.raml
>   ────────────────────────
> 
>     Node: #/web-api/https%3A%2F%2F%7BshortCode%7D.api.commercecloud.salesforce.com%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/shortCode
>       [Breaking] Parameter Added: shortCode
> 
>     Node: #/web-api/https%3A%2F%2F%7BshortCode%7D.api.commercecloud.salesforce.com%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/version
>       [Breaking] Parameter Added: version
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D/patch
>       [Breaking] Display Name Changed: patchSitesByIdOrdersById → updateOrder
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fconfirmation-status/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdConfirmationStatus → updateOrderConfirmationStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fexport-status/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdExportStatus → updateOrderExportStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fexternal-status/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdExternalStatus → updateOrderExternalStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-status/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdPaymentStatus → updateOrderPaymentStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fshipping-status/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdShippingStatus → updateOrderShippingStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fstatus/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdStatus → updateOrderStatus
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-instruments%2F%7BpaymentInstrumentId%7D/patch
>       [Breaking] Display Name Changed: patchSitesByIdOrdersByIdPaymentInstrumentsById → updateOrderPaymentInstrument
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fpayment-instruments%2F%7BpaymentInstrumentId%7D%2Ftransaction/patch
>       [Breaking] Display Name Changed: patchSitesByIdOrdersByIdPaymentInstrumentsByIdTransaction → updateOrderPaymentTransaction
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Forders%2F%7BorderNo%7D%2Fshipments%2F%7BshipmentId%7D%2Fshipping-address/put
>       [Breaking] Display Name Changed: putSitesByIdOrdersByIdShipmentsByIdShippingAddress → updateOrderShippingAddress
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.15/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
>       [Breaking] Parameter Added: WWW-Authenticate
> 
>     Node: #/web-api/https%3A%2F%2Fanypoint.mulesoft.com%2Fmocking%2Fapi%2Fv1%2Flinks%2F2a968e56-a214-422b-a6e2-61ed41faad5f%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/version
>       [Breaking] Parameter Removed: version
> 
>     Node: #/web-api/https%3A%2F%2Fanypoint.mulesoft.com%2Fmocking%2Fapi%2Fv1%2Flinks%2F2a968e56-a214-422b-a6e2-61ed41faad5f%2Fcheckout%2Forders%2F%7Bversion%7D/parameter/shortCode
>       [Breaking] Parameter Removed: shortCode
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.8/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Removed: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.8/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
>       [Breaking] Parameter Removed: WWW-Authenticate
> 
>     Summary
>     ───────
>       Breaking Changes: 19
>       Ignored Changes: 163
> 
> File: cdn-api-process-apis/cdnapi-process-api.raml
>   ──────────────────────────────────────────────────
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2Finfo/get
>       [Breaking] Display Name Changed: getZoneInfo → getZonesInfo
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Fgroups%2F%7BgroupId%7D/put
>       [Breaking] Display Name Changed: updateWafGroupById → updateWafGroup
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Fwaf%2Frules%2F%7BruleId%7D/put
>       [Breaking] Display Name Changed: updateWafRuleById → updateWafRule
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/get
>       [Breaking] Display Name Changed: getFirewallRuleById → getFirewallRule
> 
>     Node: #/web-api/end-points/%2Forganizations%2F%7BorganizationId%7D%2Fzones%2F%7BzoneId%7D%2Ffirewall%2Frules%2F%7BfirewallRuleId%7D/put
>       [Breaking] Display Name Changed: updateFirewallRuleById → updateFirewallRule
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/bearer-token.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Added: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.12/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
>       [Breaking] Parameter Added: WWW-Authenticate
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.10/SecuritySchemes/am-oauth-2.raml#/fragment/parameter/Authorization
>       [Breaking] Parameter Removed: Authorization
> 
>     Node: ./exchange_modules/893f605e-10e2-423a-bdb4-f952f56eb6d8/commerce-cloud-standards/1.0.10/SecuritySchemes/am-oauth-2.raml#/fragment/401/parameter/WWW-Authenticate
>       [Breaking] Parameter Removed: WWW-Authenticate
> 
>     Summary
>     ───────
>       Breaking Changes: 10
>       Ignored Changes: 68
> 
> APIs Added
> ──────────
>   shopper-login/shopper-login.raml
> 
> APIs Removed
> ────────────
>   shopper-stores/shopper-stores.raml
>   einstein-api-quick-start-guide/einstein-api-quick-start-guide.raml
> 
> Summary
> ───────
>   Changed APIs: 18
>     - Breaking Changes: 564
>     - Ignored Changes: 3147
>   Added APIs: 1
>   Removed APIs: 2
> ```

Looks good to me.  consider re-naming API level summary to API Summary. Just to differentiate from the overall summary

wjhsf commented:
> For the formatter code, we want to have the formatter interface that is implemented by the ConsoleFormatter. This way we can add more formatters for the same interface in future. With this design, the toString() methods written in data classes should be the methods of ConsoleFormatter class. This also separates formatting logic from the actual data.

That seems like an unnecessary abstraction to me. These classes don't implement any shared interfaces, and are unlikely to do so in the future. Additionally, having various `toString` methods on the base class is common in JavaScript (e.g. the built-in `Date` object).

unandyala commented on resources/diff/rules/defaultRules.json:
```diff
@@ -96,7 +96,7 @@
   },
   {
-    "name": "Detect addition of a required parameter",
+    "name": "Parameter Added",
```
This is addition of the required parameter. Name should be descriptive enough and indicate what the rule does exactly. This is the reason I prefer the existing names

wjhsf replied:
Thanks for pointing out that I missed a bit of important information!

I think that the old names were too long and seem more like descriptions - a name should work in multiple contexts (and ideally be short). For example, compare these two sentences:
> 3 breaking changes found: Parameter Removed, Required Parameter Added, Version Changed.

> 3 breaking changes found: Detect removal of a parameter, Detect addition of a required parameter, Detect change in API version.

I think that the first one is a lot cleaner and easier to understand.

unandyala commented on resources/diff/rules/defaultRules.json:
```diff
@@ -166,7 +166,7 @@
   },
   {
-    "name": "Ignore all changes in examples",
+    "name": "Example Changed",
```
same here- this rule ignores all changes to the example nodes

wjhsf replied:
This rule detects that an example has been changed. It is currently in the category "Ignored". None of the other rules have their category in the name, and I don't think that this should either.

unandyala replied:
my only point is, the rule name should be descriptive enough so that we/user can understand the rule with out looking at all the conditions with in the rule

wjhsf replied:
Yes, and I think that "Example Changed" is descriptive of the conditions that trigger the rule. "Ignored" isn't part of that, it's part of how we handle the rule once triggered. And I think that the name should be a label for what the rule does, not how we use the result.

To put it another way, I don't see the category as fundamentally part of the rule. If we change the category, then I don't think the rule itself has changed. If the rule hasn't changed, then the name shouldn't need to be updated. Therefore, I think that the category should not be included in the name, otherwise the name would have to be updated even when the rule has not changed.

unandyala commented on src/diff/changes/apiChanges.ts:
```diff
@@ -6,6 +6,7 @@
  */
 import { NodeChanges } from "./nodeChanges";
 import { RuleCategory } from "../ruleSet";
+import _ from "lodash";
```
Avoid using lodash as we are planning to get rid of it?

wjhsf replied:
Aww, I was hoping no one would notice so I could be lazy.

monkey-jeff replied:
I agree with this sentiment if you are using new lodash functionality.  If this is using lodash methods that we already use elsewhere in this project then don't try to fix it here,

unandyala commented on src/diff/changes/apiChanges.ts:
```diff
@@ -76,4 +84,75 @@ export class ApiChanges {
+   *
+   * @param extraIndent - Number of levels the headings should be nested
+   */
+  toString(extraIndent = 0): string {
```
I feel formatting logic should be separated from the data.  This is not a simple toString but a formatting feature. As mentioned before we need a ConsoleFormatter class with an interface so that we can add other formatters (like html) in future

monkey-jeff replied:
I completely agree and was going to make the same comment.

unandyala commented on src/diff/diffCommand.ts:
```diff
@@ -56,6 +56,12 @@ Exit statuses:
+    format: flags.enum({
+      char: "f",
+      description:
+        "Format of the output. Defaults to JSON if --out-file is specified, otherwise text.",
```
should the default be formatted text in the file? It will useful in our release process

wjhsf replied:
I was thinking that if you specify a file you probably want to use the output for something, and that JSON is more usable for programmatic purposes. ¯\\\_(ツ)_/¯

wjhsf replied:
Leaving this as is, mostly because it's late. I'm not opposed to changing it, though.

unandyala commented on src/diff/changes/apiCollectionChanges.ts:
```diff
@@ -52,4 +54,112 @@ export class ApiCollectionChanges {
+   *
+   * @param extraIndent - Number of levels the headings should be nested
+   */
+  toString(extraIndent = 0): string {
```
same here

monkey-jeff commented on src/diff/changes/apiChanges.test.ts:
```diff
@@ -10,18 +10,26 @@ import { RuleCategory } from "../ruleSet";
+  return nodeChanges;
+}
 
+function buildApiChanges(categories = [[RuleCategory.BREAKING]]): ApiChanges {
```
I find it strange that we default these functions to having changes with breaking changes.  But just a nit.

wjhsf replied:
The original function always used a breaking change. I wanted to be able to specify a different type of change without changing the original function calls.

monkey-jeff commented on src/diff/changes/apiCollectionChanges.ts:
```diff
@@ -52,4 +54,112 @@ export class ApiCollectionChanges {
+    const changedEntries = Object.entries(this.changed).filter(([, api]) => {
+      return api.hasCategorizedChanges();
+    });
+    if (changedEntries.length > 0) {
```
A lot of this logic could be simplified by using a template...something like handlebars perhaps.  This has a lot of complex data specific logic here.

wjhsf replied:
I considered using handlebars when I first started working on this. I can't even vaguely remember why I decided not to do it, it's a much better idea. :|

monkey-jeff commented on src/diff/diffCommand.ts:
```diff
@@ -161,12 +181,12 @@ Exit statuses:
     }
     if (apiChanges.hasBreakingChanges()) {
-      this.error("Breaking changes found.", { exit: 1 });
+      this.exit(1);
```
Why this change?

wjhsf replied:
The new formatted output tells you whether or not breaking changes are found, so the extra output is not necessary.

monkey-jeff reviewed (dismissed):
Really would like to see a stronger interface between the data model and the console output. Right now they are very intertwined.

wjhsf commented:
Updated to use the new formatter interface. Output is the same, except `File: #/node-id` is no longer fully underlined.

Formatters tests are not fully implemented, but I have listed all the ones I intend to add (hopefully tomorrow). Will also update docs tomorrow.

unandyala commented on src/diff/changes/apiChanges.ts:
```diff
@@ -76,4 +98,48 @@ export class ApiChanges {
+   * @param format Name of format to use
+   * @see DiffCommand for a list of supported formats
+   */
+  toFormattedString(format: string): string {
```
overall looks good compared to the previous logic. I still think data should be separated from formatter mainly for the extendibility reasons. Probably fine for now
